lib/knapsack_pro/extensions/rspec_extension.rb

```diff
@@ -0,0 +1,164 @@
+# frozen_string_literal: true
+
+module KnapsackPro
+  module Extensions
+    # Knapsack extension on top of RSpec methods.
+    #
+    # You should use methods with the `knapsack__` prefix.
+    # The methods should hide public and private RSpec methods, as well as a complex chain of internal RSpec methods.
+    # You don't break the Law of Demeter when you use a `knapsack__` method. You can use complex or chained internal RSpec methods that are hidden inside of a `knapsack__` method.
+    #
+    # Often, the structure of the `knapsack__` method is inspired by an existing RSpec method. The `knapsack__` method should have a similar structure to the original RSpec method.
+    # It should be easy to see a code difference when a new RSpec version is released.
+    # Please use permalinks to RSpec source code in your comments when you create a new `knapsack__` method.
+    module RSpecExtension
+      def self.setup!
+        RSpec::Core::World.prepend(World)
+        RSpec::Core::Runner.prepend(Runner)
+        RSpec::Core::Configuration.prepend(Configuration)
+      end
+
+      module World
+        # Based on `announce_filters`
+        # https://github.com/rspec/rspec-core/blob/f8c8880dabd8f0544a6f91d8d4c857c1bd8df903/lib/rspec/core/world.rb#L171
+        def knapsack__setup
+          fail_if_config_and_cli_options_invalid
+        end
+      end
+
+      module Runner
+        def knapsack__mark_setup_as_done
+          @knapsack__setup_done = true
+        end
+
+        def knapsack__setup_done?
+          !!@knapsack__setup_done
+        end
+
+        # Based on:
+        # https://github.com/rspec/rspec-core/blob/f8c8880dabd8f0544a6f91d8d4c857c1bd8df903/lib/rspec/core/runner.rb#L98
+        def knapsack__setup(stream_error = $stderr, stream_out = $stdout)
+          # Abstract from RSpec::Core::Runner#setup, since we do not need to set any filters or files at this point,
+          # and we do not want to let world.announce_filters be called, since it will print
+          # out `No examples found.` message.
+          configure(stream_error, stream_out)
+
+          world.knapsack__setup
```

Comment:
I wanted to have better isolation from modules so that the `def knapsack__setup(stream_error = $stderr, stream_out = $stdout)` method is not spaghetti composed of multiple methods from different modules.

For example, `World#knapsack__setup` does a simple method call: `fail_if_config_and_cli_options_invalid`, but in fact, we got rid of many lines of code from the original method: 
https://github.com/rspec/rspec-core/blob/f8c8880dabd8f0544a6f91d8d4c857c1bd8df903/lib/rspec/core/world.rb#L171

Thanks to isolation, you can find the difference between our code and the original code easier if you look at the `World#knapsack__setup` method and what it is based on.

Reply:
> you can find the difference between our code and the original code easier

In my opinion, it doesn't make it easier.

`fail_if_config_and_cli_options_invalid` does not come from `setup` (as `knapsack__setup` would suggest). Also, we added a module, that monkey patches World using a `prepend` to avoid having `fail_if_config_and_cli_options_invalid` inlined here.

If we needed to point at people to `fail_if_config_and_cli_options_invalid` we could inline and add a comment:

```ruby
# Based on:
# LINK
world.fail_if_config_and_cli_options_invalid
```

But I would also skip the comment since I can grep the RSpec codebase for `fail_if_config_and_cli_options_invalid` easily.

---

Having said that, feel free to keep the separation if you feel strongly about it.

Reply:
I can't call `world.fail_if_config_and_cli_options_invalid` because it's a private method of `World`.

I came up with something that is closer to the original RSpec code structure:
https://github.com/KnapsackPro/knapsack_pro-ruby/pull/237/commits/9e73e094a4b666a90537dab0de77d46bbb06d50f